node/pkg/p2p/cutover.go

```diff
@@ -0,0 +1,138 @@
+// If the current time is after the configured time, we should start up using the new quic-v1 bootstrap peers immediately. If the configured time is in the future,
+// we need to start up with the existing bootstrap string and cut over to the new quic-v1 at the specified time.
+func checkForCutOver(logger *zap.Logger, bootstrapPeers string, ccqBootstrapPeers string, components *Components) (newBootstrapPeers string, newCcqBootstrapPeers string, err error) {
+	newBootstrapPeers, newCcqBootstrapPeers, delay, err := checkForCutOverImpl(logger, bootstrapPeers, ccqBootstrapPeers, components, cutOverTimeStr, time.Now())
+
+	if delay != time.Duration(0) {
+		// Wait for the cut over time and then panic so we restart with the new quic-v1. TODO: Can we just restart p2p??
+		go func() {
+			time.Sleep(delay)
+			logger.Info("time to cut over to new quic-v1", zap.String("cutOverTime", cutOverTimeStr), zap.String("component", "p2pco"))
+			panic("p2pco: time to cut over to new quic-v1")
+		}()
+	}
+
+	return newBootstrapPeers, newCcqBootstrapPeers, err
+}
+
+func checkForCutOverImpl(logger *zap.Logger, bootstrapPeers string, ccqBootstrapPeers string, components *Components, coTimeStr string, now time.Time) (newBootstrapPeers string, newCcqBootstrapPeers string, delay time.Duration, err error) {
+	newBootstrapPeers = bootstrapPeers
+	newCcqBootstrapPeers = ccqBootstrapPeers
+
+	if coTimeStr == "" {
+		return
+	}
+
+	bootstrapIsV1, err := validateQuic(bootstrapPeers)
+	if err != nil {
+		logger.Error(`bootstrap peers string is invalid:`, zap.String("bootstrapPeers", bootstrapPeers), zap.Error(err), zap.String("component", "p2pco"))
+		err = fmt.Errorf("unexpected format of bootstrap peers: %w", err)
+		return
+	}
+
+	if ccqBootstrapPeers != "" {
+		ccqBootstrapIsV1, ccqerr := validateQuic(ccqBootstrapPeers)
+		if ccqerr != nil {
+			logger.Error(`ccq bootstrap peers string is invalid:`, zap.String("ccqBootstrapPeers", ccqBootstrapPeers), zap.Error(ccqerr), zap.String("component", "p2pco"))
+			err = fmt.Errorf("unexpected format of ccq bootstrap peers: %w", ccqerr)
+			return
+		}
+
+		if bootstrapIsV1 != ccqBootstrapIsV1 {
+			logger.Error(`there is a mismatch between bootstrap peers and ccq bootstrap peers:`,
+				zap.Bool("bootstrapIsV1", bootstrapIsV1),
+				zap.Bool("ccqBootstrapIsV1", ccqBootstrapIsV1),
+				zap.String("bootstrapPeers", bootstrapPeers),
+				zap.String("ccqBootstrapPeers", ccqBootstrapPeers),
+				zap.Error(err), zap.String("component", "p2pco"),
+			)
+			err = fmt.Errorf("quic version mismatch between bootstrap peers and ccq bootstrap peers")
+			return
+		}
+	}
+
+	if bootstrapIsV1 {
+		for _, la := range components.ListeningAddressesPatterns {
+			if strings.Contains(la, "quic") && !strings.Contains(la, "quic-v1") {
+				err = fmt.Errorf("bootstrapPeers has been updated to quic-v1, but components.ListeningAddressesPatterns has not: %s", la)
```

Comment:
just want to confirm my understanding of this condition, `ListeningAddressesPatterns` has not yet been changed before this point, so this would trip if a guardian / spy attempts to pass in a v1 url right now? is the thought that `ListeningAddressesPatterns` would be changed in the future (in which case this would need to be reversed?) or is `ListeningAddressesPatterns` supposed to be conditionally set before this and I'm missing something

Reply:
this might work in the tests because you provide different components, but it looks like the `p2p.go` code always uses hardcoded `/quic` ones. perhaps it would be easier to deal with the default string with a ternary there and not have the awkward side effect here.

Reply:
Reworked the design.